### PR TITLE
Copy all properties and descriptors when patching React

### PR DESCRIPTION
**Pull request by nkzawa**
Fixes https://github.com/zeit/next.js/issues/899

### CODE REVIEW

File: client/patch-react.js [hunk 1]
```diff
@@ -105,11 +106,11 @@ function wrap (fn, around) {
...
-  // copy all properties
-  Object.assign(_fn, fn)
-  _fn.displayName = fn.displayName || fn.name
-
-  _fn.prototype = fn.prototype
+  for (const [k, d] of Object.entries(getOwnPropertyDescriptors(fn))) {
+    try {
+      Object.defineProperty(_fn, k, d)
+    } catch () {}
+  }
```
Comment (nkzawa) [inline client/patch-react.js]: `name`, `displayName` and `prototype` are also copied with this.

File: client/patch-react.js [hunk 2]
```diff
@@ -105,11 +106,11 @@ function wrap (fn, around) {
     return around.call(this, fn, ...args)
   }
 
-  // copy all properties
-  Object.assign(_fn, fn)
-  _fn.displayName = fn.displayName || fn.name
```
Comment (arunoda) [inline client/patch-react.js]: I think we need to set the displayName manually.
If there's a displayName that's fine.

But if that's not, we need to use fn's className as the display name.

Comment (nkzawa) [inline client/patch-react.js]: `defineProperty` can override `name` too, we don't need to do that as far as I tested with Chrome.

**Comment by nkzawa** [comment]
removed polyfill of `getOwnPropertyDescriptors` since `babel-runtime` has it.

File: client/patch-react.js (hunk 2 above)
Comment (arunoda) [inline client/patch-react.js]: @nkzawa then that's awesome.

**Comment by sakulstra** [comment]
I tested the pr and it fixes the issues for me. Is there anything blocking a merge? Eventually one could add tests for the example apps to prevent issues like this in the future?
next.js is basically not usable from beta .18 to .26 for mobx users :)

**Comment by rauchg** [comment]
@sakulstra if you can submit a PR with said tests it'd be very welcome. Since we don't use mobx, we can't guarantee a regression on this.

**Comment by sakulstra** [comment]
I'll try to ;), thx for merging

**Pull request merged**